File: app/src/main/java/com/pomo/sync/recovery/ReplicaRecovery.kt
package com.pomo.sync.recovery

internal data class FrontierHead(val feedKey: String, val sequence: Long, val operationId: String)
internal data class CheckpointManifest(
    val checkpointId: String,
    val kind: String,
    val frontier: List<FrontierHead>,
    val materializerVersion: Int,
    val projectionRoot: String,
    val packIds: List<String>,
    val blobIds: List<String>,
)
internal data class PackedOperation(val operationId: String, val feedKey: String, val sequence: Long, val wire: ByteArray)
internal data class JournalPack(val packId: String, val prefix: FrontierHead, val operations: List<PackedOperation>)

internal object CheckpointPolicy {
    fun validate(manifest: CheckpointManifest) {
        require(manifest.kind in setOf("ROUTINE", "RECOVERY", "SAFETY"))
        require(manifest.frontier == manifest.frontier.sortedBy { it.feedKey })
        require(manifest.frontier.map { it.feedKey }.distinct().size == manifest.frontier.size)
        require(manifest.materializerVersion > 0 && manifest.projectionRoot.isNotBlank())
        require(manifest.packIds.distinct().size == manifest.packIds.size)
        require(manifest.blobIds.distinct().size == manifest.blobIds.size)
    }

    fun validatePack(pack: JournalPack, forkedFeeds: Set<String>) {
        require(pack.prefix.feedKey !in forkedFeeds)
        require(pack.operations.isNotEmpty())
        require(pack.operations.all { it.feedKey == pack.prefix.feedKey })
        require(pack.operations.map { it.sequence } == (1L..pack.prefix.sequence).toList())
        require(pack.operations.last().operationId == pack.prefix.operationId)
        require(pack.operations.map { it.operationId }.distinct().size == pack.operations.size)
    }
}

internal data class RecoverySource(val sourceId: String, val checkpoint: CheckpointManifest?, val operations: List<PackedOperation>)
internal data class RehydrationPlan(
    val checkpoint: CheckpointManifest?,
    val operations: List<PackedOperation>,
    val sourceByOperation: Map<String, Set<String>>,
    val gaps: Set<String>,
)

internal object Rehydrator {
    fun plan(sources: Collection<RecoverySource>): RehydrationPlan {
        val checkpoints = sources.mapNotNull { it.checkpoint }.onEach(CheckpointPolicy::validate)
        val checkpoint = checkpoints.maxWithOrNull(compareBy<CheckpointManifest> { it.frontier.sumOf { head -> head.sequence } }.thenBy { it.checkpointId })
        val operations = sources.flatMap { it.operations }.groupBy { it.operationId }
        operations.forEach { (_, copies) -> require(copies.map { it.wire.toList() }.distinct().size == 1) { "Operation ID collision" } }
        val selected = operations.values.map { it.first() }.sortedWith(compareBy<PackedOperation> { it.feedKey }.thenBy { it.sequence })
        val gaps = linkedSetOf<String>()
        selected.groupBy { it.feedKey }.forEach { (feed, values) ->
            val sequences = values.map { it.sequence }.toSet()
            for (sequence in 1L..(values.maxOfOrNull { it.sequence } ?: 0)) if (sequence !in sequences) gaps += "$feed@$sequence"
        }
        val provenance = operations.mapValues { (_, copies) ->
            sources.filter { source -> copies.any { copy -> source.operations.any { it.operationId == copy.operationId } } }.mapTo(linkedSetOf()) { it.sourceId }
        }
        return RehydrationPlan(checkpoint, selected, provenance, gaps)
    }
}

internal enum class IntegrityFailure { PROJECTION_CORRUPT, JOURNAL_CORRUPT, DEVICE_KEY_MISSING }
internal data class SafeModeState(val active: Boolean, val incarnationSealed: Boolean, val inspectionAllowed: Boolean, val exportAllowed: Boolean)

internal fun integrityDisposition(failure: IntegrityFailure): SafeModeState =
    when (failure) {
        IntegrityFailure.PROJECTION_CORRUPT -> SafeModeState(false, false, true, true)
        IntegrityFailure.JOURNAL_CORRUPT, IntegrityFailure.DEVICE_KEY_MISSING -> SafeModeState(true, true, true, true)
    }

internal fun quarantineThresholdExceeded(quarantined: Int, suspendedFeeds: Int): Boolean = quarantined > 1_000 || suspendedFeeds > 16

File: extension/src/sync/recovery/replicaRecovery.ts
export interface FrontierHead { readonly feedKey: string; readonly sequence: number; readonly operationId: string }
export interface CheckpointManifest {
  readonly checkpointId: string; readonly kind: "ROUTINE" | "RECOVERY" | "SAFETY"; readonly frontier: readonly FrontierHead[];
  readonly materializerVersion: number; readonly projectionRoot: string; readonly packIds: readonly string[]; readonly blobIds: readonly string[];
}
export interface PackedOperation { readonly operationId: string; readonly feedKey: string; readonly sequence: number; readonly wire: Uint8Array }
export interface JournalPack { readonly packId: string; readonly prefix: FrontierHead; readonly operations: readonly PackedOperation[] }

export function validateCheckpoint(manifest: CheckpointManifest): void {
  const sorted = [...manifest.frontier].sort((left, right) => left.feedKey.localeCompare(right.feedKey));
  if (sorted.some((entry, index) => entry.feedKey !== manifest.frontier[index]?.feedKey) || new Set(sorted.map((entry) => entry.feedKey)).size !== sorted.length) throw new Error("checkpoint frontier must be unique and sorted");
  if (!Number.isSafeInteger(manifest.materializerVersion) || manifest.materializerVersion < 1 || manifest.projectionRoot.length === 0) throw new Error("invalid checkpoint roots");
  if (new Set(manifest.packIds).size !== manifest.packIds.length || new Set(manifest.blobIds).size !== manifest.blobIds.length) throw new Error("duplicate checkpoint objects");
}

export function validateJournalPack(pack: JournalPack, forkedFeeds: ReadonlySet<string>): void {
  if (forkedFeeds.has(pack.prefix.feedKey) || pack.operations.length === 0) throw new Error("pack prefix is not replaceable");
  for (let index = 0; index < pack.operations.length; index++) {
    const operation = pack.operations[index]!;
    if (operation.feedKey !== pack.prefix.feedKey || operation.sequence !== index + 1) throw new Error("pack must be a complete feed prefix");
  }
  if (pack.operations.at(-1)?.operationId !== pack.prefix.operationId || new Set(pack.operations.map((operation) => operation.operationId)).size !== pack.operations.length) throw new Error("invalid pack head");
}

export interface RecoverySource { readonly sourceId: string; readonly checkpoint: CheckpointManifest | null; readonly operations: readonly PackedOperation[] }
export interface RehydrationPlan {
  readonly checkpoint: CheckpointManifest | null; readonly operations: readonly PackedOperation[];
  readonly sourceByOperation: ReadonlyMap<string, ReadonlySet<string>>; readonly gaps: ReadonlySet<string>;
}

export function planRehydration(sources: readonly RecoverySource[]): RehydrationPlan {
  const checkpoints = [...new Map(sources.flatMap((source) => source.checkpoint === null ? [] : [[source.checkpoint.checkpointId, source.checkpoint] as const])).values()];
  checkpoints.forEach(validateCheckpoint);
  const dominates = (left: CheckpointManifest, right: CheckpointManifest): boolean => {
    const rightHeads = new Map(right.frontier.map((head) => [head.feedKey, head]));
    return [...rightHeads.values()].every((rightHead) => {
      const leftHead = left.frontier.find((head) => head.feedKey === rightHead.feedKey);
      return leftHead !== undefined && (leftHead.sequence > rightHead.sequence ||
        (leftHead.sequence === rightHead.sequence && leftHead.operationId === rightHead.operationId));
    });
  };
  const maximal = checkpoints.filter((candidate) => checkpoints.every((other) => other === candidate || !dominates(other, candidate)));
  if (maximal.length > 1) throw new Error("recovery checkpoints are incomparable");
  checkpoints.splice(0, checkpoints.length, ...maximal);
  const copies = new Map<string, Array<{ sourceId: string; operation: PackedOperation }>>();
  for (const source of sources) for (const operation of source.operations) copies.set(operation.operationId, [...(copies.get(operation.operationId) ?? []), { sourceId: source.sourceId, operation }]);
  const operations: PackedOperation[] = [];
  const sourceByOperation = new Map<string, ReadonlySet<string>>();
  for (const [operationId, candidates] of copies) {
    const first = candidates[0]!.operation;
    if (candidates.some(({ operation }) => !equalBytes(first.wire, operation.wire))) throw new Error("Operation ID collision");
    operations.push(first);
    sourceByOperation.set(operationId, new Set(candidates.map(({ sourceId }) => sourceId)));
  }
  operations.sort((left, right) => left.feedKey.localeCompare(right.feedKey) || left.sequence - right.sequence);
  const gaps = new Set<string>();
  const feeds = new Map<string, PackedOperation[]>();
  for (const operation of operations) feeds.set(operation.feedKey, [...(feeds.get(operation.feedKey) ?? []), operation]);
  for (const [feed, values] of feeds) {
    const sequences = new Set(values.map((operation) => operation.sequence));
    for (let sequence = 1; sequence <= Math.max(...sequences); sequence++) if (!sequences.has(sequence)) gaps.add(`${feed}@${sequence}`);
  }
  return { checkpoint: checkpoints.at(-1) ?? null, operations, sourceByOperation, gaps };
}

export type IntegrityFailure = "PROJECTION_CORRUPT" | "JOURNAL_CORRUPT" | "DEVICE_KEY_MISSING";
export function integrityDisposition(failure: IntegrityFailure): { readonly active: boolean; readonly incarnationSealed: boolean; readonly inspectionAllowed: true; readonly exportAllowed: true } {
  return { active: failure !== "PROJECTION_CORRUPT", incarnationSealed: failure !== "PROJECTION_CORRUPT", inspectionAllowed: true, exportAllowed: true };
}
export function quarantineThresholdExceeded(quarantined: number, suspendedFeeds: number): boolean { return quarantined > 1_000 || suspendedFeeds > 16; }
function equalBytes(left: Uint8Array, right: Uint8Array): boolean { return left.length === right.length && left.every((value, index) => value === right[index]); }
